File: 0007-reverse-integer/0007-reverse-integer.java
class Solution {
    public int reverse(int x) {
        int ans = 0;
        boolean isNegative = false;
        if (x < 0) {
            x *= -1;
            isNegative = true;
        }
        while (x > 0) {
            if (Integer.MAX_VALUE / 10 < ans) {
                return 0;
            }
            int rem = x % 10;
            ans = ans * 10 + rem;
            x = x / 10;
        }
        return isNegative ? -1 * ans : ans;
    }
}

File: 0007-reverse-integer/0007-reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        ans, max_val = 0, 2 ** 31
        neg = -1 if x < 0 else 1
        x = abs(x)
        while x:
            rem = x % 10
            x //= 10
            if (ans > max_val / 10) or (ans == max_val // 10 and rem > max_val % 10):
                return 0
            ans  = (ans * 10) + rem
            
        return ans * neg

# Approach: Mathematical algo to revser number. To handle 32 bit constraint, check that updated ans should not be more than
# max_val (2^31) before actually making the update.